src/main/flight/imu.c

```diff
@@ -252,8 +252,21 @@ STATIC_UNIT_TESTED void imuMahonyAHRSupdate(float dt, float gx, float gy, float
 #ifdef USE_MAG
     // Use measured magnetic field vector
     fpVector3_t mag_bf = {{mag.magADC[X], mag.magADC[Y], mag.magADC[Z]}};
-    float recipMagNorm = vectorNormSquared(&mag_bf);
-    if (useMag && recipMagNorm > 0.01f) {
+    fpMat33_t rMatYTrans, rMatPR;
+    // Using negative angle yields transposed rotation matrix
```

Comment:
This is overkill if only heading is needed. Simply rotate mag to earth frame (matrixVectorMul), compute angle an add yaw from matix (-atan2_approx(rMat[1][0], rMat[0][0])).

And if full vector is needed, then rotate to EF and then rotate by  vector rMat(yaw) (instead of multiplying matrices)

Reply:
You're right. I refactored the code according to your suggestion.